firebase-firestore/src/main/java/com/google/firebase/firestore/core/SyncEngine.java

```diff
@@ -589,4 +595,19 @@ private boolean errorIsInteresting(Status error) {
 
     return false;
   }
+
+  private boolean isRetryableError(Exception e) {
+    if (e instanceof FirebaseFirestoreException) {
+      // In transactions, the backend will fail outdated reads with FAILED_PRECONDITION and
+      // non-matching document versions with ABORTED. These errors should be retried.
+      List<FirebaseFirestoreException.Code> retryableExceptions =
+          Arrays.asList(
+              FirebaseFirestoreException.Code.ABORTED,
+              FirebaseFirestoreException.Code.FAILED_PRECONDITION);
+
+      return retryableExceptions.contains(((FirebaseFirestoreException) e).getCode())
+          || !Datastore.isPermanentError(((FirebaseFirestoreException) e).getCode());
```

Comment:
I don't care strongly, but unless we expect this list to grow I'd probably just write the simplest code for this:
```
import com.google.firebase.firestore.FirebaseFirestoreException.Code;
...
    if (e instanceof FirebaseFirestoreException) {
      // In transactions, the backend will fail outdated reads with FAILED_PRECONDITION and
      // non-matching document versions with ABORTED. These errors should be retried.
      Code code = ((FirebaseFirestoreException)e).getCode();
      return code == Code.ABORTED || 
          code == Code.FAILED_PRECONDITION || 
          !Datastore.isPermanentError(code);
    }
```